module/persistent.lua

```diff
@@ -0,0 +1,290 @@
+
+    if args.create_clients == true then
+        for pid, client in pairs(self.restored_settings.clients) do
+            awful.spawn.easy_async_with_shell(string.format("ps aux | grep '%s' | grep -v 'grep'", client.command), function(stdout)
+                if stdout == "" or stdout == nil then
+                    local new_pid = awful.spawn(client.command, false)
+                    self.restored_settings.clients[tostring(new_pid)] = self.restored_settings.clients[pid]
+                    self.restored_settings.clients[pid] = nil
+                end
+            end)
+        end
+
+        gtimer {
+            timeout = 3,
+            call_now = false,
+            autostart = true,
+            single_shot = true,
+            callback = function()
+                reapply_clients(self)
+            end
+        }
+    else
+        reapply_clients(self)
+    end
+end
+
+function persistent:save(args)
+    args = args or {}
+
+    args.save_tags = args.save_tags == nil and true or args.save_tags
+    args.save_clients = args.save_clients == nil and true or args.save_clients
+
+    if args.save_tags == true then
+        save_tags(self)
+    end
+    if args.save_clients ==  true then
+        save_clients(self)
+    end
+
+    local cache = gfilesystem.get_xdg_cache_home()
+    if not gfilesystem.dir_readable(cache .. "awesome") then
+        gfilesystem.make_directories(cache .. "awesome")
+    end
+
+    local json_settings = json.encode(self.settings, { indent = true })
+    local path = cache .. "awesome/persistent.json"
+    awful.spawn.with_shell("echo '" .. json_settings .. "'" .. " > " .. path)
+end
+
+function persistent:restore(args)
+    args = args or {}
+
+    args.restore_tags = args.restore_tags == nil and true or args.restore_tags
+    args.restore_clients = args.restore_clients == nil and true or args.restore_clients
+
+    local path = gfilesystem.get_xdg_cache_home() .. "awesome/persistent.json"
+    awful.spawn.easy_async_with_shell("cat " .. path, function(stdout)
+        self.restored_settings = json.decode(stdout)
+        if self.restored_settings ~= nil then
+            if args.restore_tags == true then
+                restore_tags(self, args)
+            end
+            if args.restore_clients ==  true then
+                restore_clients(self, args)
+            end
+        end
+    end)
+end
+
+function persistent:enable(args)
+    args = args or {}
+
+    capi.awesome.connect_signal("exit", function()
+        self:save(args)
+    end)
+
+    capi.awesome.connect_signal("startup", function()
+        if is_restart() == false then
+            args.create_clients = args.create_clients == nil and true or args.create_clients
+        end
+
+        self:restore(args)
+    end)
+end
+
+local function new()
+    local ret = gobject{}
+    gtable.crush(ret, persistent, true)
+    ret.settings = {}
+
+    return ret
+end
+
+if not instance then
+    instance = new()
+end
```

Comment:
Pretty sure you are missing some return statement here? The user needs some way to interact with the module right?